simulator.py

```diff
@@ -0,0 +1,126 @@
+                             ['pnl', 'cost', 'position', 'position_np'])
+        self.__portfolio = list()
+        self.__key = uuid.uuid4()
+        self.decode_data(data)
+
+    def decode_data(self, data):
+        """
+        Decode while in multi-process.
+
+        In multi-process, simulator will get the data reference
+        which alpha need. To prevent polluteing the origin data, we need to
+        copy a local version.
+        """
+        self.data = copy.deepcopy(data)
+
+    def next_date(self):
+        """
+        Move the simulating date.
+        """
+        for dataset in self.data.values():
+            dataset.next_date(self.__key)
+        self.data['target_data'].next_date(self.__key)
+        self.data['workdays'].next_date(auth_key=self.__key)
+
+    def produce_portfolio(self, today):
+        """
+        Find daily portfolio.
+        """
+        portfolio = self.alpha_formula(today, self.data['others'])
+        portfolio.normalize()
+        portfolio.encode_to_nparray(self.settings.target)
+        self.__portfolio.append(portfolio)
+
+    def produce_pnl(self):
+        """
+        Calculate pnl for alphas.
+        This method will be called by simulation day by day while inputting
+        a new portfolio.
+
+        While calculating pnl,
+        we should take a look at our asset and trading cost.
+            (1) Total asset
+                We assume that our assets are the same everyday (equal to
+                self.initial_asset). To achieve this goal, we open a virtual
+                account to borrow/lend money to the investing account (this
+                account is self.cur_asset[self.base]).
+            (2) Trading cost
+                There are different functions about calculating trading cost
+                in different investing targets.
+                This method should import different modules while targets are
+                different.
+
+                In our template function, we calculate trading cost while
+                selling out the asset.
+                The trading_cost function could be built by alpha writers.
+                The API will adjust in the future, while the fuction needs
+                more data.
+        """
+        self.cost_object.reset()
+        today = self.__portfolio[-1]
+        target_data = self.data['target_data']
+        for instrument in today:
+            if target_data[instrument][-1].sell is not None:
+                self.cost_object(today, instrument, target_data[instrument][-1])
+        profit, cost = self.cost_object.result()
+        self.__result.append(
+            (today.date, profit - cost, cost, self.__portfolio[-1],
+             self.__portfolio[-1].position_np))
+
+    def run(self):
+        """
+        Carried out simulating process.
+        """
+        self.setting()
+        self.alpha_formula = self.settings.alpha(
+            self.settings.start_date, self.data, self.settings.parameters)
```

Comment:
Nothing other than `produce_portfolio` needs this variable. If we unfold `produce_portfolio` into this method, we won't need to save this under `self`.